File: week-6/sentimental-credit/credit.py
from cs50 import get_int


def main():
    # Get card input
    input = get_int("Number: ")
    card = str(input)

    ##### Implement Luhn's Algorithm #####
    sum = 0

    # Multiply every second number by 2
    digit = []
    if len(card) % 2 == 0:
        isEven = True
    else:
        isEven = False

    for i in range(len(card)):
        # Process cards with an even number of digits
        cur_num = int(card[i])
        if (i % 2 == 0 and isEven) or (i % 2 == 1 and not isEven):
            cur_num *= 2
            if cur_num > 9:
                cur_num = (cur_num % 10) + 1
        # Add whatever the odd digit or modified even digit is
        sum += cur_num

    # Detect valid card
    if sum % 10 == 0:
        length = len(card)
        # Detect digit diferences for card carrier
        if (card[0] == "4") and (length == 13 or length == 16):
            print("VISA")
        elif (card[0] == "3" and (card[1] == "4" or card[1] == "7")) and (length == 15):
            print("AMEX")
        elif (card[0] == "5" and (card[1] >= "1" and card[1] <= "5")) and (length == 16):
            print("MASTERCARD")
        else:
            print("INVALID")
    # Detect invalid card
    else:
        print("INVALID")


main()

File: week-1/credit/credit.c
#include <stdio.h>
#include <cs50.h>
#include <math.h>

int main(void)
{
    //get card info
    unsigned long int card;
    do
    {
        card = get_long("Card number: ");
    }
    while (card < 1);

    //prepare card info to be spliced into array
    const int CARD_LENGTH = (int) log10((double) card);

    //splice card info
    int digits[CARD_LENGTH];
    for (int i = 0; card > 0; i++)
    {
        digits[i] = card % 10;
        card /= 10;
    }


    //implement Luhn's algorithm

    //multiply every second number by 2
    int evenDigits[(int)(CARD_LENGTH / 2)];
    for (int i = 0; i <= (int)(CARD_LENGTH / 2); i ++)
    {
        evenDigits[i] = 0;
        evenDigits[i] = digits[(2 * i) + 1] * 2;
    }


    //solve issue where some 13 digit cards are handled incorrectly by c
    const int TEMP_FIRST_DIGIT = digits[CARD_LENGTH];


    //add individual digits together
    int total = 0;
    for (int i = 0; i < ((int)(CARD_LENGTH / 2) + CARD_LENGTH % 2) ; i++)
    {
        //adds 1 if over 10
        if (evenDigits[i] > 9)
        {
            total ++;
        }

        //adds ones place of evenDigits[i]
        total += evenDigits[i] % 10;

    }

    digits[CARD_LENGTH] = TEMP_FIRST_DIGIT;

    //cycle through odd digits
    for (int i = 0; i <= CARD_LENGTH; i += 2)
    {
        total += digits[i];
    }
    if (total % 10 == 0)
    {
        if (digits[CARD_LENGTH] == 3 && (digits[CARD_LENGTH - 1] == 4 || digits[CARD_LENGTH - 1] == 7))
        {
            printf("AMEX\n");
        }
        else if (digits[CARD_LENGTH] == 5 && digits[CARD_LENGTH - 1] >= 1 && digits[CARD_LENGTH - 1] <= 5)
        {
            printf("MASTERCARD\n");
        }
        else if (digits[CARD_LENGTH] == 4)
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
    return 0;


}
